xrdp/xrdp_egfx.c

```diff
@@ -0,0 +1,1021 @@
+    {
+        return 1;
+    }
+    in_uint32_le(s, queueDepth);
+    in_uint32_le(s, intframeId);
+    in_uint32_le(s, totalFramesDecoded);
+    LOG(LOG_LEVEL_DEBUG, "xrdp_egfx_process_frame_ack: queueDepth %d"
+        " intframeId %d totalFramesDecoded %d",
+        queueDepth, intframeId, totalFramesDecoded);
+    if (egfx->frame_ack != NULL)
+    {
+        egfx->frame_ack(egfx->user, queueDepth, intframeId, totalFramesDecoded);
+    }
+    return 0;
+}
+
+/******************************************************************************/
+/* RDPGFX_CMDID_CAPSADVERTISE */
+static int
+xrdp_egfx_process_capsadvertise(struct xrdp_egfx *egfx, struct stream *s)
+{
+    int index;
+    int capsSetCount;
+    int caps_count;
+    int version;
+    int capsDataLength;
+    int flags;
+    char *holdp;
+    int *versions;
+    int *flagss;
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_process_capsadvertise:");
+    if (egfx->caps_advertise == NULL)
+    {
+        return 0;
+    }
+    in_uint16_le(s, capsSetCount);
+    if ((capsSetCount < 1) || (capsSetCount > 1024))
+    {
+        return 1;
+    }
+    caps_count = 0;
+    versions = g_new(int, capsSetCount);
+    if (versions == NULL)
+    {
+        return 1;
+    }
+    flagss = g_new(int, capsSetCount);
+    if (flagss == NULL)
+    {
+        g_free(versions);
+        return 1;
+    }
+    for (index = 0; index < capsSetCount; index++)
+    {
+        if (!s_check_rem(s, 8))
+        {
+            return 1;
+        }
+        in_uint32_le(s, version);
+        in_uint32_le(s, capsDataLength);
+        if (!s_check_rem(s, capsDataLength))
+        {
+            return 1;
+        }
+        holdp = s->p;
+        if (capsDataLength == 4)
+        {
+            in_uint32_le(s, flags);
+            versions[caps_count] = version;
+            flagss[caps_count] = flags;
+            caps_count++;
+        }
+        s->p = holdp + capsDataLength;
+    }
+    if (caps_count > 0)
+    {
+        egfx->caps_advertise(egfx->user, caps_count, versions, flagss);
+    }
+    g_free(versions);
+    g_free(flagss);
+    return 0;
+}
+
+/******************************************************************************/
+static int
+xrdp_egfx_process(struct xrdp_egfx *egfx, struct stream *s)
+{
+    int error;
+    int cmdId;
+    int flags;
+    int pduLength;
+    char *holdp;
+    char *holdend;
+
+    LOG(LOG_LEVEL_TRACE, "xrdp_egfx_process:");
+    error = 0;
+    while (s_check_rem(s, 8))
+    {
+        holdp = s->p;
+        holdend = s->end;
+        in_uint16_le(s, cmdId);
+        in_uint16_le(s, flags);
+        in_uint32_le(s, pduLength);
+        s->end = holdp + pduLength;
+        LOG(LOG_LEVEL_DEBUG, "xrdp_egfx_process: cmdId 0x%x flags %d "
+            " pduLength %d", cmdId, flags, pduLength);
+        if (pduLength < 8)
+        {
+            return 1;
+        }
+        if (!s_check_rem(s, pduLength - 8))
+        {
+            return 1;
+        }
+        switch (cmdId)
```

Comment:
You must be able to handle the Cache Import Offer PDU (and reply to it), even when you don't want to import anything.

Reply:
good point

Reply:
https://github.com/neutrinolabs/xrdp/issues/2502

Reply:
> https://github.com/neutrinolabs/xrdp/issues/2502

No, that is not enough or right: Quoting the docs here (`1.5.2 Server Implementation Requirements`):

```
Furthermore, servers implementing the Remote Desktop Protocol: Graphics Pipeline Extension must
be capable of processing the following messages:
- RDPGFX_FRAME_ACKNOWLEDGE_PDU (section 2.2.2.13)
- RDPGFX_CACHE_IMPORT_OFFER_PDU (section 2.2.2.16)
- RDPGFX_CAPS_ADVERTISE_PDU (section 2.2.2.18)
```

Also, `3.2.5.16 Processing an RDPGFX_CACHE_IMPORT_OFFER_PDU message`:

```
Once the RDPGFX_CACHE_IMPORT_OFFER_PDU message has been processed, the server MUST
respond by sending the RDPGFX_CACHE_IMPORT_REPLY_PDU (section 2.2.2.17) message to the
client (section 3.2.5.17).
```

Reply:
Got it! I copied your comment into the issue so we don't lose it.

What's the best way to test this? Is there a way to force FreeRDP to send this message so we can test the negotiation? Or is there a mode on the Microsoft clients we can set to trigger it?

Reply:
mstsc sends that PDU (though not always).